File: src/test/decimal.cpp
/*
 * NPG-explorer, Nucleotide PanGenome explorer
 * Copyright (C) 2012-2015 Boris Nagaev
 *
 * See the LICENSE file for terms of use.
 */

#include <boost/test/unit_test.hpp>

#include "Decimal.hpp"

BOOST_AUTO_TEST_CASE (Decimal_main) {
    using namespace npge;
    BOOST_CHECK(D(1.01) + Decimal() == D(1.01));
    BOOST_CHECK(D(1.01) + Decimal() >= D(1.01));
    BOOST_CHECK(D(1.01) + Decimal() <= D(1.01));
    BOOST_CHECK(D(1.01) > D(1.001));
    BOOST_CHECK(D(1.001) < D(1.01));
    BOOST_CHECK(D(-1.001) < D(1.01));
    BOOST_CHECK(D(-1.001) > D(-1.01));
    BOOST_CHECK(D(0.001) > Decimal());
    BOOST_CHECK(D(0.001).to_s() == "0.001" ||
                D(0.001).to_s() == ".001");
    BOOST_CHECK(D(0).to_s() == "0.0" ||
                D(0).to_s() == "0." ||
                D(0).to_s() == ".0" ||
                D(0).to_s() == "0");
    BOOST_CHECK(Decimal(Decimal().to_s()) == Decimal());
    BOOST_CHECK(D(0.0) == Decimal());
    BOOST_CHECK(D(.0) == Decimal());
    BOOST_CHECK(D(0.) == Decimal());
    BOOST_CHECK(D(0) == Decimal());
    BOOST_CHECK(D(1) / 2 == D(0.5));
    BOOST_CHECK(D(0.75) * D(0.5) == D(0.375));
    int tenth = Decimal::sub_point / 10;
    BOOST_CHECK(D(1.3).fraction() == 3 * tenth);
    BOOST_CHECK(D(-1.3).fraction() == 3 * tenth);
    BOOST_CHECK(D(1.3).to_i() == 1);
    BOOST_CHECK(D(-1.3).to_i() == -1);
    BOOST_CHECK(D(1.3).round() == 1);
    BOOST_CHECK(D(-1.3).round() == -1);
    BOOST_CHECK(D(1.7).fraction() == 7 * tenth);
    BOOST_CHECK(D(-1.7).fraction() == 7 * tenth);
    BOOST_CHECK(D(1.7).to_i() == 1);
    BOOST_CHECK(D(-1.7).to_i() == -1);
    BOOST_CHECK(D(1.7).round() == 2);
    BOOST_CHECK(D(-1.7).round() == -2);
    BOOST_CHECK(D(0.9) * 100 == 90);
    BOOST_CHECK((D(0.9) * 100).to_i() == 90);
}

File: test-lua/decimal.lua
one = Decimal(1)
assert(one:to_s() == '1')
assert(Decimal(one) == one)
assert(Decimal(1) == Decimal('1'))
assert(Decimal(1) ~= Decimal('1.7'))
assert(Decimal(10) * Decimal('0.1') == Decimal(1))
assert(Decimal('20.1') > Decimal('20'))
assert(Decimal('20.1') >= Decimal('20'))
assert(Decimal('-20.1') < Decimal('20'))
assert(Decimal('-20.1') < Decimal('-20'))
assert(Decimal('-20.0') == Decimal('-20'))
assert(Decimal('0') == Decimal(0))
assert(Decimal('100') * Decimal(0) == Decimal(0))
assert(Decimal('1.2'):round() == 1)
assert(Decimal('1.2'):to_d() >= 1.199)
assert(Decimal('1.2'):to_d() <= 1.201)
assert(-Decimal('1.2') == Decimal('-1.2'))
assert(-Decimal() == Decimal())
assert(Decimal(0) == Decimal('0'))
assert(Decimal(0) == Decimal())
assert(Decimal('1.2') + Decimal('3.4') == Decimal('4.6'))
assert(Decimal('1.2') - Decimal('3.4') == Decimal('-2.2'))
assert(Decimal('1.2') * Decimal('3.4') == Decimal('4.08'))
assert(Decimal('1.7') / Decimal('3.4') == Decimal('0.5'))
assert(Decimal('1.7'):to_s() == '1.7')

